### PR TITLE
Remove anonymous trait params from 2018 and beyond

### PR DESCRIPTION
**Pull request by mark-i-m**
cc @Centril @nikomatsakis 
cc #41686 rust-lang/rfcs#2522 #53272 

This PR removes support for anonymous trait parameters syntactically in rust 2018 and onward.

TODO:
- [x] Add tests

### CODE REVIEW

File: src/test/ui/anon-params-denied-2018.rs [hunk 1]
```diff
@@ -1,4 +1,6 @@
-// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// Test that anonymous parameters are disallowed in 2018 edition.
```
Comment (varkor) [inline src/test/ui/anon-params-denied-2018.rs]: I think the copyright notice has to be the first thing in the file.

File: src/test/ui/anon-params-deprecated.fixed [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
```
Comment (varkor) [inline src/test/ui/anon-params-deprecated.fixed]: *2018

File: src/test/ui/anon-params-denied-2018.rs (hunk 1 above)
Comment (mark-i-m) [reply src/test/ui/anon-params-denied-2018.rs]: Ah, thanks. I didn't notice I had done that.

**Comment by mark-i-m** [comment]
Actually, if we wait for #53617, we won't need the license at all.

**Comment by mark-i-m** [comment]
Ugh... The line numbers changed. I will wait until the license isn't needed and then fix.

**Comment by petrochenkov** [comment]
r=me after fixing tests

**Comment by mark-i-m** [comment]
Tests should be fixed hopefully (this is taking a while to run locally)

**Comment by mark-i-m** [comment]
@petrochenkov time for another review? Thanks!

**Pull request merged**